builtin/providers/aws/resource_aws_instance.go

```diff
@@ -986,7 +997,15 @@ func buildAwsInstanceOpts(
 		}
 	}
 
-	if hasSubnet && associatePublicIPAddress {
+	if hasNetworkInterface {
+		ni := &ec2.InstanceNetworkInterfaceSpecification{
+			NetworkInterfaceId: aws.String(networkInterfaceID),
+			DeviceIndex:        aws.Int64(int64(0)),
+			SubnetId:           aws.String(subnetID),
```

Comment:
Hey @johnrengelman terribly sorry for the silence here.
Looks at the SDK, it seems you are not supposed to supply the subnet id if you're supplying a network interface id, so I think you're idea on not allowing the others is correct. 

Regarding validation, I honestly don't know from this vantage, we may just need to fall through to any error returned from AWS